tests/neg/EqualityStrawman1.scala

```diff
@@ -0,0 +1,79 @@
+package strawman.equality
+import annotation.implicitNotFound
+
+object EqualityStrawman1 {
+
+  trait Eq[-T]
+
+  @implicitNotFound("cannot compare value of type ${T} with a value outside its equality class")
+  trait Impossible[T]
+
+  object Eq extends Eq[Any]
+
+  trait Base {
+    def === (other: Any): Boolean = this.equals(other)
+    def === [T <: CondEquals](other: T)(implicit ce: Impossible[T]): Boolean = ???
+  }
+
+  trait CondEquals extends Base {
+    def === [T >: this.type <: CondEquals](other: T)(implicit ce: Eq[T]): Boolean = this.equals(other)
+    def === [T](other: T)(implicit ce: Impossible[T]): Boolean = ???
+  }
+
+  trait Equals[-T] extends CondEquals
```

Comment:
On Fri, Apr 22, 2016 at 5:46 PM, Guillaume Martres <notifications@github.com
> wrote:

> In tests/neg/EqualityStrawman1.scala
> <https://github.com/lampepfl/dotty/pull/1231#discussion_r60760108>:
>
> > +  @implicitNotFound("cannot compare value of type ${T} with a value outside its equality class")
> > +  trait Impossible[T]
> > +
> > +  object Eq extends Eq[Any]
> > +
> > +  trait Base {
> > +    def === (other: Any): Boolean = this.equals(other)
> > +    def === [T <: CondEquals](other: T)(implicit ce: Impossible[T]): Boolean = ???
> > +  }
> > +
> > +  trait CondEquals extends Base {
> > +    def === [T >: this.type <: CondEquals](other: T)(implicit ce: Eq[T]): Boolean = this.equals(other)
> > +    def === [T](other: T)(implicit ce: Impossible[T]): Boolean = ???
> > +  }
> > +
> > +  trait Equals[-T] extends CondEquals
>
> Why have both CondEquals and Equals[-T]? Do they serve different purpose?
>

No. The next PR has only one class: EqClass for both.

 - Martin


> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/1231/files/89f9091c2d10aa23c5d6259694df104fc14c60f7#r60760108>
>



-- 
Prof. Martin Odersky
LAMP/IC, EPFL